src/service/mod.rs

```diff
@@ -0,0 +1,72 @@
+pub mod contact_service;
+
+use std::sync::Arc;
+
+use rocket::{http::Status, response::Responder};
+use thiserror::Error;
+
+use super::{dht::Client, Config};
+use crate::persistence;
+use crate::persistence::FileBasedContactStore;
+use contact_service::{ContactService, ContactServiceApi};
+
+/// Generic result type
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Generic error type
+#[derive(Debug, Error)]
+pub enum Error {
+    /// all errors originating from the persistence layer
+    #[error("Persistence error: {0}")]
+    Persistence(#[from] persistence::Error),
+
+    /// errors that currently return early http status code Status::NotAcceptable
+    #[error("not acceptable")]
+    PreconditionFailed,
+}
+
+/// Map from service errors directly to rocket status codes. This allows us to
+/// write handlers that return `Result<T, service::Error>` and still return the correct
+/// status code.
+impl<'r, 'o: 'r> Responder<'r, 'o> for Error {
+    fn respond_to(self, req: &rocket::Request) -> rocket::response::Result<'o> {
+        match self {
+            // for now handle all persistence errors as InternalServerError, there
+            // will be cases where we want to handle them differently (eg. 409 Conflict)
+            Error::Persistence(_) => Status::InternalServerError.respond_to(req),
+            Error::PreconditionFailed => Status::NotAcceptable.respond_to(req),
+        }
+    }
+}
+
+/// A depencency container for all services that are used by the application
+#[derive(Clone)]
+pub struct ServiceContext {
+    pub config: Config,
```

Comment:
Since `Config` is set at the start of the application and never changes, we could use something like `once_cell::lazy`, or `lazy_static`, to initialize it once as static and have it available as a global throughout the codebase.

Something like

```
lazy_static! {
    // Globally accessible configuration
    static ref CFG: config::Config =
        Config::parse();
}
```

That way we would never have to clone it, or worry about lifetime issues in general.

Do you think this would make sense API-wise? (We can create a separate task for the actual change, if we want to do it)

Reply:
Hey, I did not know that there is a macro for that looks a lot cleaner that way. Yes I like the idea of lazy constants for smaller values but again for testing it is sometimes useful to run with a different configuration. If you access the config from global you would have to change env vars for test runs to get the same effect. But we can definitely do a lazy init for the conf anyhow we should keep it as a constructor arg in context though.

Reply:
Created backlog ticket for this here: https://github.com/BitcreditProtocol/E-Bill/issues/192